packages/react-devtools-shared/src/utils.js

```diff
@@ -305,3 +323,277 @@ export function setInObject(
+//   [
+//     123,
+//     "abc",
+//     [true, false],
+//     { foo: 123, bar: "abc" }
+//   ];
+//
+// Would show a preview of...
+//   [123, "abc", Array(2), {…}]
+export function formatDataForPreview(
+  data: any,
+  showFormattedValue: boolean,
+): string {
+  if (data != null && data.hasOwnProperty(meta.type)) {
+    return showFormattedValue
+      ? data[meta.preview_long]
+      : data[meta.preview_short];
+  }
+
+  const type = getDataType(data);
+
+  switch (type) {
+    case 'html_element':
+      return `<${truncateForDisplay(data.tagName.toLowerCase())} />`;
+    case 'function':
+      return truncateForDisplay(data.name);
+    case 'string':
+      return `"${data}"`;
+    case 'bigint':
+      return truncateForDisplay(data.toString() + 'n');
+    case 'symbol':
+      return truncateForDisplay(data.toString());
+    case 'react_element':
+      return `<${truncateForDisplay(
+        getDisplayNameForReactElement(data) || 'Unknown',
+      )} />`;
+    case 'array_buffer':
+      return `ArrayBuffer(${data.byteLength})`;
+    case 'data_view':
+      return `DataView(${data.buffer.byteLength})`;
+    case 'array':
+      if (showFormattedValue) {
+        const formatted = [];
+        for (let i = 0; i < data.length; i++) {
+          formatted[i] = formatDataForPreview(data[i], false);
+        }
+        return `[${truncateForDisplay(formatted.join(', '))}]`;
+      } else {
+        const length = data.hasOwnProperty(meta.size)
+          ? data[meta.size]
+          : data.length;
+        return `Array(${length})`;
+      }
+    case 'typed_array':
+      const shortName = `${data.constructor.name}(${data.length})`;
+      if (showFormattedValue) {
+        return `${shortName} [${truncateForDisplay(
+          data
+            .toString()
+            .split(',')
+            .join(', '),
+        )}]`;
+      } else {
+        return shortName;
+      }
+    case 'iterator':
+      const name = data.constructor.name;
+      if (showFormattedValue) {
+        const valueStrings = [];
+        // TRICKY
+        // Don't use [...spread] syntax for this purpose.
+        // This project uses @babel/plugin-transform-spread in "loose" mode which only works with Array values.
+        // Other types (e.g. typed arrays, Sets) will not spread correctly.
+        Array.from(data).forEach(entryOrEntries => {
+          // TRICKY
+          // Browsers display Maps and Sets differently.
+          // To mimic their behavior, detect if we've been given an entries tuple.
+          //   Map(2) {"abc" => 123, "def" => 123}
+          //   Set(2) {"abc", 123}
+          if (Array.isArray(entryOrEntries)) {
+            const key = formatDataForPreview(entryOrEntries[0], true);
+            const value = formatDataForPreview(entryOrEntries[1], false);
+            valueStrings.push(`${key} => ${value}`);
+          } else {
+            valueStrings.push(formatDataForPreview(entryOrEntries, false));
+          }
```

Comment:
Same question. Should we break down earlier if we know we’re gonna go over the truncation limit anyway, and work for next items is wasted?